### PR TITLE
WIP: Replace Javascript with Typescript

### LINKED ISSUE #40: Replace Javascript front-end code with Typescript
The front-end of the electron app is written in vanilla javascript and jquery.

It's rather ugly but it works. Many people advised that rewritting in Typescript will help improve it and make it more maintainable.

### PR DESCRIPTION
**Pull request by kelsos**
Closes  #40 

@LefterisJP is the title enough or can you put some kind of work in progress label?

### CODE REVIEW

**Comment by LefterisJP** [comment]
@kelsos title is enough. But I can create a work in progress label too. Won't hurt.

**Comment by kelsos** [comment]
Small status update. I still need to figure out the packaging of the compiled Typescript files, otherwise the debug version seems to work.

**Comment by LefterisJP** [comment]
What needs to be done for the packaging of the compiled typescript files? You mean to create a release, right? Doesn't electron packager somehow take care of it.

**Comment by kelsos** [comment]
Yes they should be packaged properly, but I was putting the result to the wrong folder and the application failed to start. Also I should probably add the `npm run build` step to the package.sh before the packaging.

**Comment by LefterisJP** [comment]
Please try it. If you need any help or need me to test anything please tell me.

**Comment by kelsos** [comment]
@LefterisJP Could you please review the code, check for anything I might have broken and verify that the packaged version works for you properly?

**Comment by kelsos** [comment]
Yes, please just give me some time as it will require some manual picking

File: package.json [hunk 1]
```diff
@@ -11,16 +11,30 @@
...
+        "@types/datatables.net": "^1.10.13",
+        "@types/jquery": "^3.3.6",
+        "@types/jquery.contextmenu": "^1.7.33",
+        "@types/mocha": "^5.2.5",
+        "@types/node": "^8.9.2",
+        "@types/tail": "^1.2.0",
         "chai": "^4.1.2",
         "chai-as-promised": "^7.1.1",
-        "electron": "1.8.4",
+        "electron": "^1.8.4",
```
Comment (LefterisJP) [inline package.json]: Can you leave it to `1.8.4` for now? As discussed I wanna specifically update electron alone at some point (soon). But until then it should be pinned to a single version.

File: package.json [hunk 2]
```diff
@@ -11,16 +11,30 @@
...
+        "rimraf": "^2.6.2",
+        "spectron": "^4.0.0",
+        "ts-node": "^7.0.1",
+        "tslint": "^5.11.0",
+        "typescript": "^3.0.3"
     },
     "scripts": {
-        "test": "mocha ui/tests/",
-        "start": "electron ."
+        "build": "rimraf ui/dist && tsc",
```
Comment (LefterisJP) [inline package.json]: Why rimraf and not just simply `rm -rf`. What is the advantage? I see the package but it does not explain why: https://github.com/isaacs/rimraf

File: tslint.json [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    "static-method",
+                    "instance-method"
+                ]
+            }
+        ],
+        "no-arg": true,
+        "no-bitwise": true,
+        "no-console": [
+            true,
+//            "log",
```
Comment (LefterisJP) [inline tslint.json]: is this commented out? What does it do? Can't you just remove it?

File: ui/rotkehlchen_service.ts [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+export class NoResponseError extends Error {
+    constructor() {
+        super('No response received from rpc');
+    }
+}
+
+export class NoPremiumCredentials extends Error {
+}
+
+export class RotkehlchenService {
```
Comment (LefterisJP) [inline ui/rotkehlchen_service.ts]: nice touch

File: ui/accounting_settings.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
+import { showError, showInfo } from './utils';
```
Comment (LefterisJP) [inline ui/accounting_settings.ts]: Is the space after `{` and before `}` a typescript style?

File: ui/balances_table.ts [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+        const amount = parseFloat(query_result[asset].amount.toString());
+        const value = parseFloat(query_result[asset].usd_value.toString());
+        data[asset] = {amount: amount, usd_value: value};
+    }
+    SAVED_BALANCES[location] = data;
+    total_table_recreate();
+}
+
+function balance_table_init_callback(sett?: DataTables.Api, _json?: any) {
```
Comment (LefterisJP) [inline ui/balances_table.ts]: why `sett` and not `settings` here? `sett` is a terrible name. Why did you rename it?

File: ui/enums/BalanceStatus.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export enum BalanceStatus {
```
Comment (LefterisJP) [inline ui/enums/BalanceStatus.ts]: Why enums in their own files and directory? Isn't that an overkill?

File: ui/index.html [hunk 8]
```diff
@@ -27,19 +27,19 @@
     <!-- Data tables css -->
     <link href="../node_modules/datatables.net-dt/css/jquery.dataTables.css" rel="stylesheet" type="text/css">
 
-    <!-- Jquery ui date picker css -->
-    <link href="css/jquery.datetimepicker.min.css" rel="stylesheet" type="text/css">
+	<!-- Jquery ui date picker css -->
```
Comment (LefterisJP) [inline ui/index.html]: Ah no no no no. Please don't add tabs. Undo the indentation changes in this file.

In rotkehlchen we only use 4 spaces for indentation.

File: ui/model/action-result.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+export interface ActionResult<T> {
```
Comment (LefterisJP) [inline ui/model/action-result.ts]: That's interesting. So that's some kind of template. Even though it's called an interface. Interesting. Seems like a good idea to use this for api results.

File: ui/model/asset-balance.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export interface AssetBalance {
```
Comment (LefterisJP) [inline ui/model/asset-balance.ts]: Why have all those under a `model/` directory? I don't mind. Makes sense, to have all classed modeling objects or interfaces in a collected place. Just wondering how come you did that. Is it a standard thing to do?

File: ui/typings/JQuery.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
+interface JQuery {
```
Comment (LefterisJP) [inline ui/typings/JQuery.ts]: Does the directory **need** to be called `typings`? If not can we rename it to `typing` ?

File: ui/typings/JQuery.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    setType(type: string): void;
+
+    setTitle(title: string): void;
+
+    setContentAppend(content: string): void;
+}
+
+interface PulsateOptions {
+    color?: string;
+    reach?: number;                              // how far the pulse goes in px
```
Comment (LefterisJP) [inline ui/typings/JQuery.ts]: nitpick: Can you make the comments all start at the same line and be aligned to each other?

File: ui/utils.ts [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+
+export function unsuggest_element(selector: string) {
+    $(selector).pulsate('destroy');
+    $(selector).removeAttr('style');
+}
+
+export function suggest_element(selector: string, state_to_set: string) {
+    settings.start_suggestion = state_to_set;
+    $(selector).pulsate({
+        color: '#e45325', // set the color of the pulse
```
Comment (LefterisJP) [inline ui/utils.ts]: can you align comments here too?

File: ui/monitor.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    // else it means we still need to have data to load
+    $('#top-loading-icon').removeClass().addClass('fa fa-circle-o-notch fa-spin fa-fw');
+
+    for (const task_id in tasks_map) {
+        if (!tasks_map.hasOwnProperty(task_id)) {
+            continue;
+        }
+        const task = tasks_map[task_id];
+        if (task.id == null) {
+            console.log(`NULL TASK ID: ${JSON.stringify(task, null, 4)}`);
```
[the remark this answers is not in the document]
Comment (kelsos) [reply ui/monitor.ts]: OK, could you give me en example of what was expected here so I can try to debug it?

File: ui/dashboard.ts [hunk 15]
```diff
@@ -247,7 +247,7 @@ ipcRenderer.on('failed', () => {
 
 export function init_dashboard() {
     // add callbacks for dashboard to the monitor
-    monitor_add_callback('query_exchange_balances', (result: ExchangeBalanceResult) => {
+    monitor_add_callback('query_exchange_balances_async', (result: ExchangeBalanceResult) => {
```
Comment (kelsos) [inline ui/dashboard.ts]: Oh yeah I forgot about this sorry.

File: package.json (hunk 2 above)
Comment (kelsos) [reply package.json]: My guess, and my reason for not using `rm -rf` is that it would not work on Windows, unless you had some extra stuff installed.

File: tslint.json (hunk 3 above)
Comment (kelsos) [reply tslint.json]: I yeas sorry, I forgot it, I will do so. There are way to many console.log entries to have the linter warn about them.

File: ui/accounting_settings.ts (hunk 5 above)
Comment (kelsos) [reply ui/accounting_settings.ts]: It is configurable, you can either have it or not, I just picked some stuff from the webui rules.
If you don't like it I can change the rule and fix it.

File: ui/dashboard.ts (hunk 15 above)
Comment (LefterisJP) [reply ui/dashboard.ts]: no worries, I am still testing. Will take some time till I am sure it works fine.

File: ui/balances_table.ts (hunk 6 above)
Comment (kelsos) [reply ui/balances_table.ts]: I had a warning due to settings being already a globally declared constant. 
I will change it to something else :)

File: ui/index.html (hunk 8 above)
Comment (kelsos) [reply ui/index.html]: Ok that is weird it should not happen. I have the editor configured to use spaces. I will have a look.

File: ui/enums/BalanceStatus.ts (hunk 7 above)
Comment (kelsos) [reply ui/enums/BalanceStatus.ts]: I just didn't want to have the files in the root ui directory.

File: ui/model/action-result.ts (hunk 9 above)
Comment (kelsos) [reply ui/model/action-result.ts]: It is a generic interface. It made sense since we had a lot of common structure and it didn't make any sense creating multiple classes to model them. Instead we can only model the payload :)

File: ui/typings/JQuery.ts (hunk 11 above)
Comment (kelsos) [reply ui/typings/JQuery.ts]: I really thought I would have more of them in the folder. I will rename it :)

File: ui/model/asset-balance.ts (hunk 10 above)
Comment (kelsos) [reply ui/model/asset-balance.ts]: I am not really sure. In angular we usually have some directories to separate structure of code based on functionality, so there is a folder for components/services/models so it made sense to me to have these files in their own directory.

File: ui/dashboard.ts (hunk 15 above)
Comment (kelsos) [reply ui/dashboard.ts]: It makes sense it is a huge change. I am almost certain I broke something somewhere.

Comment (LefterisJP) [reply ui/dashboard.ts]: yep deleting editing otc trades also does not work. Will try to see why.

Comment (LefterisJP) [reply ui/dashboard.ts]: I am getting:

```
otctrades.ts:55 Uncaught TypeError: Cannot read property 'id' of undefined
    at delete_otc_trade (otctrades.ts:55)
    at jQuery.fn.init.callback (utils.ts:210)
    at HTMLLIElement.itemClick (jquery.contextMenu.min.js:1)
    at HTMLDocument.dispatch (/home/lefteris/w/rotkehlchen/node_modules/jquery/dist/jquery.js:5183)
    at HTMLDocument.elemData.handle (/home/lefteris/w/rotkehlchen/node_modules/jquery/dist/jquery.js:4991)
```

In both edit and delete seems that `const data = row.data() as OtcTrade;` returns undefined.

Why would that be happening?

Comment (kelsos) [reply ui/dashboard.ts]: I am not sure, let me have a look. Maybe it binds another object.

File: ui/utils.ts [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        ctx_menu_items['sep1'] = '---------';
+        ctx_menu_items['quit'] = {name: 'Quit', icon: 'fa-sign-out'};
+
+        // idea taken from:
+        // https://stackoverflow.com/questions/43161236/how-to-show-edit-and-delete-buttons-on-datatables-when-right-click-to-rows
+        $.contextMenu({
+            selector: `#${id}_body tr td`,
+            callback: (key: string) => {
+                // @ts-ignore
+                const tr = $(this).closest('tr');
```
Comment (kelsos) [inline ui/utils.ts]: This was the reason for the otc trade edit not working. `this` here actually refers to the module created by the `utils.ts` file so `$(this)` doesn't make actual sense.

**Comment by LefterisJP** [comment]
So fixed one more error with the CSV exporting. But I now got a problem with clicking to go to a specific exchange's page. Will leave a comment in the code directly.

File: ui/dashboard.ts [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+import {ExchangeBalanceResult} from './model/exchange-balance-result';
+import {BlockchainBalances} from './model/blockchain-balances';
+import {ipcRenderer, remote} from 'electron';
+import {service} from './rotkehlchen_service';
+import {Currency} from './model/currency';
+
+function add_exchange_on_click() {
+    $('.panel a').click((event: JQuery.Event) => {
+        event.preventDefault();
+        const target = event.target as HTMLAnchorElement;
```
Comment (LefterisJP) [inline ui/dashboard.ts]: So @kelsos this place here always returns undefined. Doing a print of the event itself I get the following:

```js
XX----> {
    "originalEvent": {
        "isTrusted": true
    },
    "type": "click",
    "target": {},
    "currentTarget": {
        "jQuery331031681506678370841": {
            "events": {
                "click": [
                    {
                        "type": "click",
                        "origType": "click",
                        "data": null,
                        "guid": 118,
                        "namespace": ""
                    }
                ]
            }
        }
    },
    "relatedTarget": null,
    "timeStamp": 17746.65,
    "jQuery33103168150667837084": true,
    "delegateTarget": {
        "jQuery331031681506678370841": {
            "events": {
                "click": [
                    {
                        "type": "click",
                        "origType": "click",
                        "data": null,
                        "guid": 118,
                        "namespace": ""
                    }
                ]
            }
        }
    },
    "handleObj": {
        "type": "click",
        "origType": "click",
        "data": null,
        "guid": 118,
        "namespace": ""
    },
    "data": null
}
```
